### PR TITLE
INTYGFV-16290: Implemented logic to handle update of issueing unit for certificate.

### PR DESCRIPTION
**Pull request by ChrilleSigerhed**

### CODE REVIEW

File: certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/CertificateDataConfigConverter.java [hunk 1]
```diff
@@ -11,19 +14,20 @@
...
+    if (DATE.equals(elementSpecification.configuration().type())) {
+      final var configuration = (ElementConfigurationDate) elementSpecification.configuration();
+      return CertificateDataConfigDate.builder()
+          .id(configuration.id())
+          .text(configuration.name())
+          .minDate(configuration.minDate())
+          .maxDate(configuration.maxDate())
+          .build();
+    }
+    return null;
```
Comment (mww-sopra-steria) [inline certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/CertificateDataConfigConverter.java:31]: If it reaches this I think it is better that we throw an exception than return null. For example, IllegalStateException which a message saying what type of configuration that didn't match an alternative. If it comes to this we have a bug in the code before and it is good to fail fast here.

File: certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/CertificateDataValueConverter.java [hunk 2]
```diff
@@ -7,22 +7,21 @@
...
-    };
+    if (ElementType.DATE.equals(elementSpecification.configuration().type())) {
+      final var value = elementValue != null ? ((ElementValueDate) elementValue).date() : null;
+      final var configuration = (ElementConfigurationDate) elementSpecification.configuration();
+      return CertificateDataValueDate.builder()
+          .id(configuration.id())
+          .date(value)
+          .build();
+    }
+    return null;
```
Comment (mww-sopra-steria) [inline certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/CertificateDataValueConverter.java:25]: If it reaches this I think it is better that we throw an exception than return null. For example, IllegalStateException which a message saying what type of configuration that didn't match an alternative. If it comes to this we have a bug in the code before and it is good to fail fast here.

File: certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/CertificateMetaDataUnitConverter.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+import se.inera.intyg.certificateservice.domain.certificate.model.ElementData;
+import se.inera.intyg.certificateservice.domain.certificate.model.ElementValueIssuingUnit;
+import se.inera.intyg.certificateservice.domain.certificate.model.IssuingUnit;
+import se.inera.intyg.certificateservice.domain.certificatemodel.model.ElementId;
+
+@Component
+public class CertificateMetaDataUnitConverter {
+
+
+  public UnitDTO convert(IssuingUnit issuingUnit, List<ElementData> elementData) {
```
Comment (mww-sopra-steria) [inline certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/CertificateMetaDataUnitConverter.java:18]: See comment on the class calling this. If elementData can be an optional you can simplify the code and just do a .map(...) to create a UnitDTO from the elementData orElse(...) create it from the IssuingUnit.

File: certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/ElementMetaDataConverter.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import se.inera.intyg.certificateservice.domain.certificate.model.ElementValueIssuingUnit;
+import se.inera.intyg.certificateservice.domain.certificatemodel.model.ElementId;
+
+@Component
+public class ElementMetaDataConverter {
+
+  public ElementData convert(UnitDTO unit) {
+    return
+        ElementData.builder()
+            .id(new ElementId(ISSUING_UNIT))
```
Comment (mww-sopra-steria) [inline certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/ElementMetaDataConverter.java:17]: See comment on the CertificateConstant class and change accordingly

**Review by mww-sopra-steria — CHANGES_REQUESTED** [review]
Nice! It works like a charm!

See comments for some feedback to make the implementation a bit more robust and maybe easier to read.

File: certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/common/CertificateConstants.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package se.inera.intyg.certificateservice.application.common;
+
+public class CertificateConstants {
+
+  private CertificateConstants() {
+    throw new IllegalStateException("Utility class");
+  }
+
+  public static final String ISSUING_UNIT = "ISSUING_UNIT";
```
[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/common/CertificateConstants.java:9]: Great idea, done!

File: certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/CertificateConverter.java [hunk 6]
```diff
@@ -46,31 +48,12 @@ public CertificateDTO convert(Certificate certificate, List<ResourceLinkDTO> res
...
-                            certificate.certificateMetaData().issuingUnit().contactInfo().email()
-                        )
-                        .isInactive(
-                            certificate.certificateMetaData().issuingUnit().inactive().value()
-                        )
-                        .build()
+                    certificateMetaDataUnitConverter.convert(
+                        certificate.certificateMetaData().issuingUnit(),
+                        certificate.elementData().stream()
+                            .filter(data -> data.value() instanceof ElementValueIssuingUnit)
```
[the remark this answers is not in the document]
Comment (ChrilleSigerhed) [reply certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/CertificateConverter.java:54]: Great suggestion! Done

File: certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/CertificateDataConfigConverter.java (hunk 1 above)
Comment (ChrilleSigerhed) [reply certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/CertificateDataConfigConverter.java:31]: Makes sense, done!

File: certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/CertificateMetaDataUnitConverter.java (hunk 3 above)
Comment (ChrilleSigerhed) [reply certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/CertificateMetaDataUnitConverter.java:18]: Thats brilliant, great suggestion!

**Review by mww-sopra-steria — APPROVED** [review]
Nice!

**Pull request merged**